src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -0,0 +1,313 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy
+import onnx
+from transformers import AutoTokenizer
+
+from deepsparse import Pipeline
+from deepsparse.engine import Context
+from deepsparse.transformers.utils.decoder_kv_cache import DecoderKVCache
+from deepsparse.transformers.utils.helpers import generate_session_id, softmax
+from sparsezoo.utils.onnx import save_onnx
+
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = ["NLDecoderEngine"]
+
+_CACHE_INPUT_NAME = "past_key_values"
+
+
+# TODO: We may need appropriate engine properties
+# so that this `Engine` has all the attributes required
+# by the pipeline
+class NLDecoderEngine:
+    """
+    The NLDecoderEngine (NaturalLanguageDecoderEngine) handles the
+    logic around the inference for Natural Language pipeline,
+    including batching and kv cache logic.
+
+    :param onnx_file_path: The path to the onnx model file
+    :param engine_type: The type of engine to use for the inference
+    :param engine_args: The arguments to pass to the engine
+    :param sequence_length: The maximum sequence length to run the engine for
+    :param multitoken: Whether to run the engine in multitoken mode
+    :param engine_context: The context to run the engine in
+    :param sampling_temperature: The temperature to use for sampling
+    :param deterministic: Whether to use deterministic sampling
+    :param tokenizer: The tokenizer to used for engine inputs
+    :param engine_context: The context to run the engine in
+    """
+
+    def __init__(
+        self,
+        onnx_file_path: str,
+        engine_type: str,
+        engine_args: Dict[str, Any],
+        # TODO: having a sequence_length as an argument
+        # to the engine is not ideal. The engine should
+        # not need to know about it, but it is required
+        # for determining whether kv cache is present or not
+        sequence_length: int,
+        multitoken: bool,
+        tokenizer: AutoTokenizer,
+        sampling_temperature: float = 1.0,
+        deterministic: bool = True,
+        engine_context: Optional[Context] = None,
+    ):
+
+        onnx_file_path, kv_cache_enabled = self.overwrite_onnx_model_inputs(
```

Comment:
self.engine -> for prompt processing (prefill=True) as well as autoregressive (prefill = False) [multitoken=False]

self.multitoken_engine -> for prompt processing (prefill=True, "autoregressive" at sequence len 64)

so 'multitoken' attribute becomes obsolete